ledger/common/utils/split.go

```diff
@@ -4,55 +4,71 @@ import (
 	"github.com/onflow/flow-go/ledger"
 )
 
-// SplitByPath splits an slice of payloads based on the value of bit (bitIndex) of paths
-func SplitByPath(paths []ledger.Path, payloads []ledger.Payload, bitIndex int) ([]ledger.Path, []ledger.Payload, []ledger.Path, []ledger.Payload) {
-	rpaths := make([]ledger.Path, 0, len(paths))
-	rpayloads := make([]ledger.Payload, 0, len(payloads))
-	lpaths := make([]ledger.Path, 0, len(paths))
-	lpayloads := make([]ledger.Payload, 0, len(payloads))
-
-	for i, path := range paths {
+// SplitByPath permutes the input paths to be partitioned into 2 parts. The first part contains paths with a zero bit
+// at the input bitIndex, the second part contains paths with a one at the bitIndex. The index of partition
+// is returned.
+// The same permutation is applied to the payloads slice.
+//
+// This would be the partition step of an ascending quick sort of paths (lexicographic order)
+// with the pivot being the path with all zeros and 1 at bitIndex.
+// The comparison of paths is only based on the bit at bitIndex, the function therefore assumes all paths have
+// equal bits from 0 to bitIndex-1
+//
+//  For instance, if `paths` contains the following 3 paths, and bitIndex is `1`:
+//  [[0,0,1,1], [0,1,0,1], [0,0,0,1]]
+//  then `SplitByPath` returns 1 and updates `paths` into:
+//  [[0,0,1,1], [0,0,0,1], [0,1,0,1]]
+func SplitByPath(paths []ledger.Path, payloads []ledger.Payload, bitIndex int) int {
+	i := 0
+	for j, path := range paths {
 		bit := Bit(path, bitIndex)
-		if bit == 1 {
-			rpaths = append(rpaths, path)
-			rpayloads = append(rpayloads, payloads[i])
-		} else {
-			lpaths = append(lpaths, path)
-			lpayloads = append(lpayloads, payloads[i])
+		if bit == 0 {
+			paths[i], paths[j] = paths[j], paths[i]
+			payloads[i], payloads[j] = payloads[j], payloads[i]
+			i++
```

Comment:
are we sure this works for input like this? 
[0000, 1000, 0000, 1000, 0000, 1000, 0000 ]

Reply:
shall we add a randomized tests to confirm the correctness

Reply:
What `bitIndex` should I look at in the example input to split? 
A randomized input is a good idea, but how can we check the computation is correct without a reference implementation? 🤔

Reply:
>how can we check the computation is correct without a reference implementation? 🤔

You can checkout create random inputs, save the output of the previous version, and use it to verify the new version

Reply:
If I save the output of the previous version, this means the inputs are rather hardcoded, but not randomized at each test run? Is that what you meant? Or did you mean leaving the previous implementation in the test files and use it to check the results are consistent ?

Reply:
In my comments, I was referring to a reference implementation of `Update` rather than `SplitPaths`. 
I also assumed the correctness of `Update` would imply the correctness of `SplitPaths` with a high probability since `SplitPaths` is called at every level of the `Update` recursions. 

I'll add a specific test for `SplitPaths`.

Reply:
Test for `SplitPaths` added https://github.com/onflow/flow-go/pull/505/commits/74c1908789984e8170680a8f2d5ac0b0982225b6